Code/Engine/GameEngine/StateMachine/Implementation/StateMachine.cpp

```diff
@@ -0,0 +1,462 @@
+
+  ezUInt32 uiStateIndex = 0;
+  if (m_pDescription->m_StateNameToIndexTable.TryGetValue(sStateName, uiStateIndex))
+  {
+    SetStateInternal(uiStateIndex);
+    return EZ_SUCCESS;
+  }
+
+  return EZ_FAILURE;
+}
+
+ezResult ezStateMachineInstance::SetState(ezUInt32 uiStateIndex)
+{
+  EZ_ASSERT_DEV(m_pDescription != nullptr, "Must have a description to set state by index");
+
+  if (uiStateIndex < m_pDescription->m_States.GetCount())
+  {
+    SetStateInternal(uiStateIndex);
+    return EZ_SUCCESS;
+  }
+
+  return EZ_FAILURE;
+}
+
+ezResult ezStateMachineInstance::SetStateOrFallback(const ezHashedString& sStateName, ezUInt32 uiFallbackStateIndex /*= 0*/)
+{
+  if (SetState(sStateName).Failed())
+  {
+    return SetState(uiFallbackStateIndex);
+  }
+
+  return EZ_SUCCESS;
+}
+
+void ezStateMachineInstance::Update()
+{
+  ezUInt32 uiNewStateIndex = FindNewStateToTransitionTo();
+  if (uiNewStateIndex != ezInvalidIndex)
+  {
+    SetState(uiNewStateIndex).IgnoreResult();
+  }
+
+  if (m_pCurrentState != nullptr)
+  {
+    void* pInstanceData = GetCurrentStateInstanceData();
+    m_pCurrentState->Update(*this, pInstanceData);
+  }
+}
+
+void ezStateMachineInstance::SetBlackboard(const ezSharedPtr<ezBlackboard>& blackboard)
+{
+  m_pBlackboard = blackboard;
+}
+
+void ezStateMachineInstance::SetStateInternal(ezUInt32 uiStateIndex)
+{
+  if (m_uiCurrentStateIndex == uiStateIndex)
+    return;
+
+  ExitCurrentState();
+
+  const auto& stateContext = m_pDescription->m_States[uiStateIndex];
+
+  m_pCurrentState = stateContext.m_pState.Borrow();
+  m_uiCurrentStateIndex = uiStateIndex;
+  m_pCurrentTransitions = &stateContext.m_Transitions;
+
+  EnterCurrentState();
+}
+
+void ezStateMachineInstance::EnterCurrentState()
+{
+  if (m_pCurrentState != nullptr)
+  {
+    void* pInstanceData = GetCurrentStateInstanceData();
+    m_pCurrentState->OnEnter(*this, pInstanceData);
+  }
+}
+
+void ezStateMachineInstance::ExitCurrentState()
+{
+  if (m_pCurrentState != nullptr)
+  {
+    void* pInstanceData = GetCurrentStateInstanceData();
+    m_pCurrentState->OnExit(*this, pInstanceData);
+  }
+}
+
+ezUInt32 ezStateMachineInstance::FindNewStateToTransitionTo()
+{
+  if (m_pDescription != nullptr)
+  {
+    for (auto& transitionContext : m_pDescription->m_FromAnyTransitions)
+    {
+      void* pInstanceData = GetInstanceData(transitionContext.m_uiInstanceDataOffset);
+      if (transitionContext.m_pTransition->IsConditionMet(*this, pInstanceData))
+      {
+        return transitionContext.m_uiToStateIndex;
+      }
+    }
+  }
+
+  if (m_pCurrentTransitions != nullptr)
+  {
+    for (auto& transitionContext : *m_pCurrentTransitions)
+    {
+      void* pInstanceData = GetInstanceData(transitionContext.m_uiInstanceDataOffset);
+      if (transitionContext.m_pTransition->IsConditionMet(*this, pInstanceData))
+      {
+        return transitionContext.m_uiToStateIndex;
+      }
+    }
+  }
```

Comment:
I would swap the two. The transitions on the current state always should have precedence over the any-transitions.